File: credit.c
#include <stdio.h>
#include <cs50.h>


bool checksum(long card);
int length(long card);
int digits(int num, long card);


int main(void)
{
    //Getting user's card's number
    long card = get_long("Credit Card Number:");

    //Checking if the card is valid
    if (checksum(card))
    {
        //Check if the card is AMEX
        if (length(card) == 15 && (digits(2, card) == 34 || digits(2, card) == 37))
        {
            printf("AMEX\n");
        }

        //Check if the card is MASTERCARD
        else if (length(card) == 16 && (digits(2, card) >= 51 && digits(2, card) <= 55))
        {
            printf("MASTERCARD\n");
        }
        //Check if the card is VISA
        else if ((length(card) == 13 || length(card) == 16) && digits(1, card) == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}


//Checksum function
bool checksum(long card)
{
    int total = 0;

    //Multipling every other digit by 2, and adding them to total.
    for (long i = card / 10; i > 0; i /= 100)
    {
        long n = i % 10 * 2;
        if (n >= 10)
        {
            total += n / 10;
            total += n % 10;
        }
        else
        {
            total += n;
        }
    }

    //Adding the sum of the digits that weren’t multiplied by 2 to total
    for (long i = card; i > 0; i /= 100)
    {
        total += i % 10;
    }

    //Check if card is valid
    if (total % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

//Card's length function
int length(long card)
{
    int length = 0;
    do
    {
        length++;
        card /= 10;
    }
    while (card > 0);

    return length;
}

//First two digits function
int digits(int num, long card)
{
    for (int i = length(card) - num; i > 0; i--)
    {
        card /= 10;
    }
    return card;
}

File: credit.py
from cs50 import get_int


# Main function
def main():
    # Get card number
    card = get_int("Number: ")

    # Check if card is valid
    if (checksum(card)):
        # Check if card is AMEX
        if len(str(card)) == 15 and (int(str(card)[:2]) == 34 or int(str(card)[:2]) == 37):
            print("AMEX")

        # Check if card is MASTERCARD
        elif len(str(card)) == 16 and (int(str(card)[:2]) >= 51 and int(str(card)[:2]) <= 55):
            print("MASTERCARD")

        # Check if card is VISA
        elif (len(str(card)) == 13 or len(str(card)) == 16) and int(str(card)[0]) % 10 == 4:
            print("VISA")

        else:
            print("INVALID")
    else:
        print("INVALID")


# Checksum function
def checksum(card):
    total = 0

    # Count total
    while (card > 0):
        total += card % 10
        card //= 10
        if (card == 0):
            break
        n = (card % 10) * 2
        total += (n % 10) + (n // 10)
        card //= 10

    # Check if total % 10 equals 0
    if (total % 10 == 0):
        return True
    return False


# Call main function
if __name__ == "__main__":
    main()
